### ISSUE TITLE
L044 unexpected exception, L045 CTE not used?

### ISSUE THREAD

**Issue by sti0** [sqlfluff/sqlfluff#2362]
<!--If this is a parsing or linting issue, please include a minimal SQL example which reproduces the issue, along with the `sqlfluff parse` output, `sqlfluff lint` output and `sqlfluff fix` output when relevant.-->

## Expected Behaviour

1. There should be no unexpected exception with L044
2. Both CTE are used. L045 should not raise.

## Observed Behaviour

```
CRITICAL   [L044] Applying rule L044 threw an Exception:                                                                                                                                                                                                              
Traceback (most recent call last):
  File "/home/vscode/workspace/_venv/sqlfluff_venv/lib/python3.8/site-packages/sqlfluff/core/rules/base.py", line 504, in crawl
    res = self._eval(
  File "/home/vscode/workspace/_venv/sqlfluff_venv/lib/python3.8/site-packages/sqlfluff/rules/L044.py", line 136, in _eval
    return self._analyze_result_columns(crawler.query_tree)
  File "/home/vscode/workspace/_venv/sqlfluff_venv/lib/python3.8/site-packages/sqlfluff/rules/L044.py", line 85, in _analyze_result_columns
    assert query.selectables
AssertionError
L:   1 | P:   1 | L044 | Unexpected exception: ; Could you open an issue at
                       | https://github.com/sqlfluff/sqlfluff/issues ? You can
                       | ignore this exception for now, by adding '--noqa: L044'
                       | at the end of line 1
L:   1 | P:   1 | L045 | Query defines a CTE (common-table expression) but does
                       | not use it.
```

## Steps to Reproduce
```sql
WITH base AS (
    SELECT
        c1 AS cx,
        c2
    FROM
        schm.tab
),
txt AS (
    VALUES
    (1,
        'v1'),
    (2,
        'v2')
    AS t
    (id,
        txt)
)
SELECT
    cb.cx,
    cb.c2,
    txt_de.txt AS de_txt
FROM
    base AS cb
LEFT JOIN txt AS txt_de ON
    cb.c2 = txt_de.id;
```

## Dialect
Exasol / Ansi

## Version
Include the output of `sqlfluff --version` along with your Python version.
If you are using dbt then please additionally include the dbt version.

sqlfluff v0.9.1
python v3.8.12


## Configuration
```
Include your SQLFluff configuration here
```

**Comment by barrywhart** [comment]
@sti0: Both these issues occur because of how the `txt` CTE is defined, using `VALUES`, which seems to be an EXASOL-specific feature. In order to handle this, we need to understand what it does. Can you provide a link to some documentation that explains it?

**Comment by sti0** [comment]
Sadly its not really good mentioned in the [docs](https://docs.exasol.com/sql/select.htm?Highlight=values_table). You could search on that site for `values_table` . 

I found another documentation from [AWS databricks](https://docs.databricks.com/sql/language-manual/sql-ref-syntax-qry-select-values.html) or even [Postgres](https://www.postgresql.org/docs/8.2/sql-values.html) which shows the functionality more clearly (so it's not a EXASOL-specific feature, even the ANSI dialect could parse this). 

Basicly you could generate a table of hardcoded values like you could when you do something like this:

```sql
SELECT 'A' AS c1, 123 AS c2, 'Hello' AS c3 FROM dual
union
SELECT 'B' AS c1, 234 AS c2, 'Hallo' AS c3 FROM dual
....
;
```

**Comment by barrywhart** [comment]
Nice, thanks! The "Examples" section of the [Postgres docs page](https://www.postgresql.org/docs/8.2/sql-values.html) you linked is really helpful.

**Comment by barrywhart** [comment]
I'm thinking we'll fix this by enhancing the `SelectCrawler` class (which is used in the implementation of both these rules) to interpret `VALUES` clauses approximately like the equivalent `SELECT` query. This could be a medium amount of work, but nothing too tricky, I think.

**Comment by barrywhart** [comment]
Note to self: This is a much simpler test case.
```
WITH txt AS (
    VALUES (1)
    AS t (id)
)

SELECT *
FROM txt
```

**Comment by johanjvrens** [comment]
This is also a valid test case that should pass but gives multiple errors including parsing errors as can be seen from this link: [sqlfluff online](https://online.sqlfluff.com/fluffed?sql=H4sIANvB6mEC_wv3DPFQKFHQSM7PiTfUUQBRRpoKjsEKGrxcCkAQ5ugT6hoMYWuoG1gEeUYYqOsoGOiZGJtaWfmF-roGeTprmOgoGGtq8nIBES9XsKuPq3OIghYvl1uQv69CCQAQwyBKYwAAAA%3D%3D&dialect=postgres)

```
WITH t (col_1, col_2) AS (
    VALUES
    ('08RIX0', 0.435::NUMERIC(4, 3))
)

SELECT *
FROM t
```

**Comment by barrywhart** [comment]
@johanjvrens: If it's not parsing, please create a separate issue for that. Thanks.

**Comment by jpy-git** [comment]
N.B. @johanjvrens' example is valid Postgres (but not Snowflake)

**Comment by jpy-git** [comment]
So on further investigation @johanjvrens' parsing errors are just because the parser doesn't like the shorthand typecast `::NUMERIC(4, 3)` in VALUES atm.

Removing that it parses fine. He is correct that it raises catastrophic erros on L044, pretty much the exact same as this issue. So this isn't just an EXASOL issue.
![image](https://user-images.githubusercontent.com/80432516/150570639-6db3321c-9aab-41ce-a6c8-53433cd08fa1.png)
